Flaky TestHelmInstall: installed helm version mismatch: expected "v3.21.3", got "v3.21.2"

The test get the latest version from the same url used by get_helm.sh, so we expect it to install this version, but it install the previous version (v3.21.2).

## TestHelmInstall/Install

Build: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/minikube/23228/pull-minikube-none-docker-linux-x86/2075608150731395072

```
{Failed  === RUN   TestHelmInstall/Install
I0710 16:08:00.004768   13728 exec_runner.go:51] Run: sudo rm -f /usr/bin/helm
I0710 16:08:00.017565   13728 helm.go:108] Installing helm latest version
I0710 16:08:00.017620   13728 exec_runner.go:51] Run: sudo bash -o errexit -c "
		curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
		chmod 700 get_helm.sh
		HELM_INSTALL_DIR=/usr/bin ./get_helm.sh"
I0710 16:08:01.428078   13728 exec_runner.go:84] Completed: sudo bash -o errexit -c "
		curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
		chmod 700 get_helm.sh
		HELM_INSTALL_DIR=/usr/bin ./get_helm.sh": (1.410427871s)
I0710 16:08:01.428117   13728 exec_runner.go:51] Run: /usr/bin/helm version --template {{.Version}}
    helm_test.go:105: installed helm version mismatch: expected "v3.21.3", got "v3.21.2"
    --- FAIL: TestHelmInstall/Install (1.47s)
}
```

## TestHelmInstall

In this build all 3 tests failed.

Build: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/minikube/23228/pull-minikube-kvm-docker-linux-x86/2075608149749927936

### Install

```
{Failed  === RUN   TestHelmInstall/Install
I0710 16:28:18.162019   12741 ssh_runner.go:191] Run: sudo rm -f /usr/bin/helm
I0710 16:28:18.181297   12741 helm.go:108] Installing helm latest version
I0710 16:28:18.181374   12741 ssh_runner.go:191] Run: sudo bash -o errexit -c "
		curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
		chmod 700 get_helm.sh
		HELM_INSTALL_DIR=/usr/bin ./get_helm.sh"
    helm_test.go:105: installed helm version mismatch: expected "v3.21.3", got "v3.21.2"
    --- FAIL: TestHelmInstall/Install (1.52s)
```

### Upgrade

Looks like same error - we know about v3.21.3, but get_helm.sh see v3.21.2.

```
{Failed  === RUN   TestHelmInstall/Upgrade
I0710 16:28:19.677653   12741 helm.go:105] Installing helm version v3.12.0
I0710 16:28:19.677709   12741 ssh_runner.go:191] Run: sudo bash -o errexit -c "
		curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
		chmod 700 get_helm.sh
		HELM_INSTALL_DIR=/usr/bin ./get_helm.sh --version v3.12.0"
    helm_test.go:137: upgraded helm version mismatch: expected "v3.21.3", got "v3.21.2"
    --- FAIL: TestHelmInstall/Upgrade (3.99s)
```

### NoChange

Seem like the same issue.

```
{Failed  === RUN   TestHelmInstall/NoChange
I0710 16:28:23.670493   12741 helm.go:108] Installing helm latest version
I0710 16:28:23.670556   12741 ssh_runner.go:191] Run: sudo bash -o errexit -c "
		curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
		chmod 700 get_helm.sh
		HELM_INSTALL_DIR=/usr/bin ./get_helm.sh"
I0710 16:28:23.953561   12741 ssh_runner.go:191] Run: /usr/bin/helm version --template {{.Version}}
I0710 16:28:24.020804   12741 helm.go:108] Installing helm latest version
I0710 16:28:24.020890   12741 ssh_runner.go:191] Run: sudo bash -o errexit -c "
		curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
		chmod 700 get_helm.sh
		HELM_INSTALL_DIR=/usr/bin ./get_helm.sh"
I0710 16:28:25.323147   12741 ssh_runner.go:231] Completed: sudo bash -o errexit -c "
		curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3
		chmod 700 get_helm.sh
		HELM_INSTALL_DIR=/usr/bin ./get_helm.sh": (1.30223122s)
I0710 16:28:25.323193   12741 ssh_runner.go:191] Run: /usr/bin/helm version --template {{.Version}}
    helm_test.go:169: helm version changed after second InstallHelm call: first "v3.21.2", second "v3.21.3"
    --- FAIL: TestHelmInstall/NoChange (1.71s)
```

## Good builds

In the same run we had 6 other good builds:
- https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/minikube/23228/pull-minikube-kvm-containerd-linux-x86/2075608149989003264
- https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/minikube/23228/pull-minikube-kvm-crio-linux-x86/2075608150106443776
- https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/minikube/23228/pull-minikube-none-containerd-linux-x86/2075608150542651392

We also had 4 expected skips with docker driver (no network access).

## Observations

- We are missing logs
  - latest version found
  - version install for every InstallHelm command. We verify the version but we don't log it. Logging it for every command will make this easier to debug
- Check that we check errors for all helm commands - logging is not consistent - maybe cause by unchecked errors
- Logging is very bad - parallel tests are logging to the same file without a prefix to identify the relevant test. When running tests in parallel we ned to log to per-test log file or logger. This requires infra change.

@Roslaan001 can you check this?

--- a/test/integration/helm_test.go
+++ b/test/integration/helm_test.go
@@ -20,19 +20,26 @@ package integration
 
 import (
 	"context"
-	"io"
-	"net/http"
 	"os/exec"
-	"strings"
 	"testing"
 	"time"
 
+	"golang.org/x/mod/semver"
 	"k8s.io/minikube/pkg/addons"
 	"k8s.io/minikube/pkg/minikube/config"
 	"k8s.io/minikube/pkg/minikube/machine"
 	_ "k8s.io/minikube/pkg/minikube/registry/drvs"
 	"k8s.io/minikube/pkg/minikube/run"
 )
+
+// minExpectedHelmVersion is the minimum helm version we expect to be installed.
+// We cannot check for the exact latest version because during Helm release rollouts,
+// CDN propagation delays can cause our version query and the get_helm.sh install script
+// to see different versions, leading to flaky test failures.
+// See https://github.com/kubernetes/minikube/issues/23323
+// Bumped occasionally (not on every release) to catch the CDN serving something
+// wildly stale or broken, without needing to track the exact current latest.
+const minExpectedHelmVersion = "v3.20.0"
 
 // TestHelmInstall integration test verifies helm installation, upgrade, and no-change behavior inside a live guest VM.
 // We test this flow because installing the latest helm allows self-healing and updating the cluster when Helm is missing, outdated, or broken.
@@ -83,8 +90,6 @@ func TestHelmInstall(t *testing.T) {
 		t.Skip("skipping: guest VM/container has no outbound internet access (this is required to download helm): https://github.com/kubernetes/minikube/issues/23275")
 	}
 
-	latestVersion := getLatestHelmVersion(t)
-
 	// 1. Install test
 	// The minikube ISO and kicbase images already come with Helm pre-installed.
 	// We delete the Helm binary here to simulate a broken cluster (e.g. if a user manually deleted or corrupted it).
@@ -101,25 +106,27 @@ func TestHelmInstall(t *testing.T) {
 		}
 
 		version := addons.HelmVersion(runner)
-		if version != latestVersion {
-			t.Fatalf("installed helm version mismatch: expected %q, got %q", latestVersion, version)
+		t.Logf("installed helm version: %s", version)
+		if semver.Compare(version, minExpectedHelmVersion) < 0 {
+			t.Fatalf("installed helm version %q is older than minimum expected %q", version, minExpectedHelmVersion)
 		}
 	})
 
 	// 2. Upgrade test
-	// This test ensures that if an older version of Helm is installed at /usr/bin/helm,
+	// This test ensures that if minExpectedHelmVersion is installed at /usr/bin/helm,
 	// calling InstallHelm with HelmOptions{} (defaults to latest) will correctly upgrade it to the latest version.
 	t.Run("Upgrade", func(t *testing.T) {
-		// Install an older Helm version (e.g. v3.12.0) directly to /usr/bin/helm
-		err := addons.InstallHelm(runner, addons.HelmOptions{Version: "v3.12.0"})
+		// Install minExpectedHelmVersion directly to /usr/bin/helm
+		err := addons.InstallHelm(runner, addons.HelmOptions{Version: minExpectedHelmVersion})
 		if err != nil {
-			t.Fatalf("failed to install older helm: %v", err)
+			t.Fatalf("failed to install helm %s: %v", minExpectedHelmVersion, err)
 		}
 
-		// Verify the older helm was installed and is executable
+		// Verify the pinned helm version was installed and is executable
 		versionOld := addons.HelmVersion(runner)
-		if versionOld != "v3.12.0" {
-			t.Fatalf("older helm version mismatch: expected \"v3.12.0\", got %q", versionOld)
+		t.Logf("installed helm version: %s (expected %s)", versionOld, minExpectedHelmVersion)
+		if versionOld != minExpectedHelmVersion {
+			t.Fatalf("helm version mismatch: expected %q, got %q", minExpectedHelmVersion, versionOld)
 		}
 
 		// Run InstallHelm, which should download and install the latest helm into /usr/bin/helm
@@ -130,11 +137,12 @@ func TestHelmInstall(t *testing.T) {
 
 		// Verify that a newer version of helm has been installed in /usr/bin/helm
 		versionNew := addons.HelmVersion(runner)
+		t.Logf("upgraded helm version: %s (from %s)", versionNew, versionOld)
 		if versionNew == versionOld {
 			t.Fatalf("helm version was not upgraded: still %q", versionNew)
 		}
-		if versionNew != latestVersion {
-			t.Fatalf("upgraded helm version mismatch: expected %q, got %q", latestVersion, versionNew)
+		if semver.Compare(versionNew, minExpectedHelmVersion) < 0 {
+			t.Fatalf("upgraded helm version %q is older than minimum expected %q", versionNew, minExpectedHelmVersion)
 		}
 	})
 
@@ -142,26 +150,28 @@ func TestHelmInstall(t *testing.T) {
 	// This test verifies that if Helm is already installed in /usr/bin/helm,
 	// running InstallHelm again is a no-op and does not modify or reinstall it.
 	t.Run("NoChange", func(t *testing.T) {
-		// Run InstallHelm to ensure latest helm is present
-		err := addons.InstallHelm(runner, addons.HelmOptions{})
+		// Run InstallHelm to ensure the pinned helm version is present
+		err := addons.InstallHelm(runner, addons.HelmOptions{Version: minExpectedHelmVersion})
 		if err != nil {
 			t.Fatalf("InstallHelm failed: %v", err)
 		}
 
 		// Retrieve current helm version details
 		firstVersion := addons.HelmVersion(runner)
+		t.Logf("first helm version: %s", firstVersion)
 		if firstVersion == "" {
 			t.Fatalf("helm not found at /usr/bin/helm")
 		}
 
-		// Run InstallHelm again
-		err = addons.InstallHelm(runner, addons.HelmOptions{})
+		// Run InstallHelm again with the same pinned version
+		err = addons.InstallHelm(runner, addons.HelmOptions{Version: minExpectedHelmVersion})
 		if err != nil {
 			t.Fatalf("InstallHelm second call failed: %v", err)
 		}
 
 		// Retrieve helm version details again and compare
 		secondVersion := addons.HelmVersion(runner)
+		t.Logf("second helm version: %s", secondVersion)
 		if secondVersion == "" {
 			t.Fatalf("helm not found at /usr/bin/helm")
 		}
@@ -169,37 +179,4 @@ func TestHelmInstall(t *testing.T) {
 			t.Fatalf("helm version changed after second InstallHelm call: first %q, second %q", firstVersion, secondVersion)
 		}
 	})
-}
-
-// getLatestHelmVersion queries the official Helm 3 latest version URL (used by get_helm.sh)
-// to find the latest Helm 3 release tag.
-func getLatestHelmVersion(t *testing.T) string {
-	t.Helper()
-	ctx, cancel := context.WithTimeout(t.Context(), 10*time.Second)
-	defer cancel()
-
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://get.helm.sh/helm3-latest-version", nil)
-	if err != nil {
-		t.Fatalf("failed to create request for latest helm version: %v", err)
-		return ""
-	}
-
-	resp, err := http.DefaultClient.Do(req)
-	if err != nil {
-		t.Fatalf("failed to query latest helm release tag: %v", err)
-		return ""
-	}
-	defer resp.Body.Close()
-
-	if resp.StatusCode != http.StatusOK {
-		t.Fatalf("unexpected status code querying latest helm release tag: %d", resp.StatusCode)
-		return ""
-	}
-
-	body, err := io.ReadAll(resp.Body)
-	if err != nil {
-		t.Fatalf("failed to read response body: %v", err)
-		return ""
-	}
-	return strings.TrimSpace(string(body))
 }

